src/app/3d/qgs3dmeasuredialog.cpp

```diff
@@ -0,0 +1,385 @@
+/***************************************************************************
+  qgs3dmeasuredialog.cpp
+  --------------------------------------
+  Date                 : Jun 2019
+  Copyright            : (C) 2019 by Ismail Sunni
+  Email                : imajimatika at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QCloseEvent>
+#include <QPushButton>
+
+#include "qgsproject.h"
+#include "qgs3dmeasuredialog.h"
+#include "qgs3dmaptoolmeasureline.h"
+#include "qdebug.h"
+#include "qgsmapcanvas.h"
+#include "qgisapp.h"
+
+Qgs3DMeasureDialog::Qgs3DMeasureDialog( Qgs3DMapToolMeasureLine *tool, Qt::WindowFlags f )
+  : QDialog( tool->canvas()->topLevelWidget(), f )
+  , mTool( tool )
+{
+  setupUi( this );
+
+  // New button
+  QPushButton *newButton = new QPushButton( tr( "&New" ) );
+  buttonBox->addButton( newButton, QDialogButtonBox::ActionRole );
+  connect( newButton, &QAbstractButton::clicked, this, &Qgs3DMeasureDialog::restart );
+
+  // Only support for cartesian
+  mCartesian->setChecked( true );
+
+  // Hide ellipsoidal and cartesian radio button
+  mCartesian->hide();
+  mEllipsoidal->hide();
+
+  // Initialize unit combo box
+  // Add a configuration button
+  QPushButton *cb = new QPushButton( tr( "&Configuration" ) );
+  buttonBox->addButton( cb, QDialogButtonBox::ActionRole );
+  connect( cb, &QAbstractButton::clicked, this, &Qgs3DMeasureDialog::openConfigTab );
+
+  repopulateComboBoxUnits();
+
+  // Choose unit
+  if ( mUseMapUnits )
+    mUnitsCombo->setCurrentIndex( mUnitsCombo->findData( QgsUnitTypes::DistanceUnknownUnit ) );
+  else
+    mUnitsCombo->setCurrentIndex( mUnitsCombo->findData( QgsProject::instance()->distanceUnits() ) );
+
+  QgsMapCanvas *canvas2D = QgisApp::instance()->mapCanvas();
```

Comment:
I think we should not work with 2D map canvas here - only focus on 3D map canvas - it has its own CRS.

Also elsewhere in this dialog I think we should not deal with 2D map canvas.

Reply:
I agree, it should be separated. But, if possible, it would be nice to get 3D length on 2D map canvas too (and fix https://github.com/qgis/QGIS/issues/25576)

Reply:
Handling 3D length on 2D canvas is much more tricky I think. In 2D canvas the user may not be aware that the lengths could be calculated in 3D and that Z coordinates may (or may not) be picked from the snapped points. On the other hand in 3D canvas it is easier because it is clear that measurement points are from 3D world _and_ the Z values are consistently applied (some layers may have relative Z values, other may have absolute Z values - all that is made consistent in the 3D scene config and 3D renderer config of individual layers).

Reply:
done